src/tree/updater_quantile_hist.cc

```diff
@@ -102,33 +104,82 @@ void QuantileHistMaker::Builder::SyncHistograms(
   builder_monitor_.Start("SyncHistograms");
 
   const bool isDistributed = rabit::IsDistributed();
-
   const size_t nbins = hist_builder_.GetNumBins();
-  common::BlockedSpace2d space(nodes_for_explicit_hist_build_.size(), [&](size_t node) {
-    return nbins;
-  }, 1024);
+  if (!isDistributed) {
+    common::BlockedSpace2d space(nodes_for_explicit_hist_build_.size(), [&](size_t node) {
+      return nbins;
+    }, 1024);
+
+    common::ParallelFor2d(space, this->nthread_, [&](size_t node, common::Range1d r) {
+      const auto entry = nodes_for_explicit_hist_build_[node];
+      auto this_hist = hist_[entry.nid];
+      // Merging histograms from each thread into once
+      hist_buffer_.ReduceHist(node, r.begin(), r.end());
+
+      if (!(*p_tree)[entry.nid].IsRoot() && entry.sibling_nid > -1 /*&& !isDistributed*/) {
+        auto parent_hist = hist_[(*p_tree)[entry.nid].Parent()];
+        auto sibling_hist = hist_[entry.sibling_nid];
+        SubtractionHist(sibling_hist, parent_hist, this_hist, r.begin(), r.end());
+      }
+    });
+  } else {
+    common::BlockedSpace2d space(nodes_for_explicit_hist_build_.size(), [&](size_t node) {
+      return nbins;
+    }, 1024);
```

Comment:
Space is the same with line 109. So you can move this out of "if - else"

Reply:
done